x-pack/filebeat/input/http_endpoint/handler.go

```diff
@@ -119,29 +140,68 @@ func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 				break
 			} else if !errors.Is(err, errNotCRC) {
 				h.metrics.apiErrors.Add(1)
-				h.sendAPIErrorResponse(w, r, h.log, http.StatusBadRequest, err)
+				h.sendAPIErrorResponse(txID, w, r, h.log, http.StatusBadRequest, err)
 				return
 			}
 		}
 
-		if err = h.publishEvent(obj, headers); err != nil {
+		acker.Add()
+		if err = h.publishEvent(obj, headers, acker); err != nil {
 			h.metrics.apiErrors.Add(1)
-			h.sendAPIErrorResponse(w, r, h.log, http.StatusInternalServerError, err)
+			h.sendAPIErrorResponse(txID, w, r, h.log, http.StatusInternalServerError, err)
 			return
 		}
 		h.metrics.eventsPublished.Add(1)
 		respCode, respBody = h.responseCode, h.responseBody
 	}
 
-	h.sendResponse(w, respCode, respBody)
-	if h.reqLogger != nil {
-		h.logRequest(r, respCode, nil)
+	acker.Ready()
+	if acked == nil {
+		h.sendResponse(w, respCode, respBody)
+	} else {
+		select {
+		case <-acked:
+			h.log.Debugw("request acked", "tx_id", txID)
+			if !timeout.Stop() {
+				<-timeout.C
+			}
+			h.sendResponse(w, respCode, respBody)
+		case <-timeout.C:
+			h.log.Debugw("request timed out", "tx_id", txID)
+			h.sendAPIErrorResponse(txID, w, r, h.log, http.StatusGatewayTimeout, errTookTooLong)
+		case <-h.ctx.Done():
+			h.log.Debugw("request context cancelled", "tx_id", txID)
+			h.sendAPIErrorResponse(txID, w, r, h.log, http.StatusGatewayTimeout, h.ctx.Err())
+		}
+		if h.reqLogger != nil {
+			h.logRequest(txID, r, respCode, nil)
+		}
 	}
 	h.metrics.batchProcessingTime.Update(time.Since(start).Nanoseconds())
 	h.metrics.batchesPublished.Add(1)
 }
 
-func (h *handler) sendAPIErrorResponse(w http.ResponseWriter, r *http.Request, log *logp.Logger, status int, apiError error) {
+var errTookTooLong = errors.New("could not publish event within timeout")
+
+func getTimeoutWait(u *url.URL, log *logp.Logger) time.Duration {
+	p := u.Query().Get("wait_for_completion_timeout")
+	if p == "" {
+		return 0
+	}
+	log.Debugw("wait_for_completion_timeout parameter", "value", p)
+	t, err := time.ParseDuration(p)
+	if err != nil {
+		log.Warnw("could not parse wait_for_completion_timeout parameter", "error", err)
+		return 0
+	}
```

Comment:
I'm wondering if this is the best behaviour for failure in parsing the timeout.

As an user of the API, if I set `wait_for_completion_timeout` I'm expecting the request to block until all events are acked, however if set a invalid value, let's say `wait_for_completion_timeout=1min`, then the request returns immediately and succeeds, which makes the user believe all events were acked immediately. 

I believe we could either:
* return an error to the user
* wait for ever (or a maximum documented timeout)

That way we do our best to fulfil the caller's request of waiting until all events are acked.


What do you think?

Reply:
I'm not sure. I think you might be right. I think that perhaps we should be properly strict here:
- confirm that there is either no query param, or exactly one query param that is `wait_for_completion_timeout` and that has a syntactically correct `time.Duration` (only checking the syntax and not the param names is a vector for typos in the param name).
- otherwise respond with a 400 and the error message explaining the issue.